gnmi_server/server.go

```diff
@@ -515,3 +527,63 @@ func (s *Server) Capabilities(ctx context.Context, req *gnmipb.CapabilityRequest
 		GNMIVersion:        "0.7.0",
 		Extension:          exts}, nil
 }
+
+type uint128 struct {
+	High uint64
+	Low  uint64
+}
+
+func (lh *uint128) Compare(rh *uint128) int {
+	if rh == nil {
+		// For MA disabled case, EID supposed to be 0.
+		rh = &uint128{High: 0, Low: 0}
+	}
+	if lh.High > rh.High {
+		return 1
+	}
+	if lh.High < rh.High {
+		return -1
+	}
+	if lh.Low > rh.Low {
+		return 1
+	}
+	if lh.Low < rh.Low {
+		return -1
+	}
+	return 0
+}
+
+// ReqFromMasterEnabledMA returns true if the request is sent by the master
+// controller.
+func ReqFromMasterEnabledMA(req *gnmipb.SetRequest, masterEID *uint128) bool {
+	// Read the election_id.
+	reqEID := uint128{High: 0, Low: 0}
+	// It can be one of many extensions, so iterate through them to find it.
+	for _, e := range req.GetExtension() {
+		ma := e.GetMasterArbitration()
+		if ma == nil {
+			continue
+		}
+		// The Master Arbitration descriptor has been found.
+		// If MA is enabled, the reqEID needs to be set to 0
+		if ma.ElectionId != nil {
+			reqEID = uint128{High: ma.ElectionId.High, Low: ma.ElectionId.Low}
+		}
+		// Use the election ID that is in the last extension, so, no 'break' here.
+	}
+	switch masterEID.Compare(&reqEID) {
```

Comment:
Since `reqEID` was inited with {0,0} value, this check fails when the current request does not have any MA extension but previous request had set one. But the [spec](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-master-arbitration.md#32-setrequest-handling) wants servers to access such requests:

> If there is no arbitration extension associated with the Set request, the Set request should be accepted and processed. Therefore, client that expects to ever be part of a replica-group MUST always set this extension, and the role MUST be populated.

Reply:
Thanks. I changed the check here. If the MA extension does not exist, EID will not be set.